File: rust/examples/secure_implementation/secure_example.rs
#[path = "../basic_implementation/schema.rs"]
//Load the schema file
mod schema;

use base64::engine::general_purpose::STANDARD as BASE64_STANDARD;
use base64::Engine;
use dotenv::from_filename;
use skypydb::{json_map, ReactiveClient, Result, SkypydbError};

fn main() -> Result<()> {
    // Load environment variables from .env.local file
    let _ = from_filename(".env.local");

    // Load encryption key from environment. Fail if not found
    let encryption_key = std::env::var("ENCRYPTION_KEY")
        .ok_or_else(|| {
            SkypydbError::validation("Missing ENCRYPTION_KEY. Set it in environment or .env.local.")
        })?;

    // Load salt key from environment. Fail if not found
    let salt_encoded = std::env::var("SALT_KEY")
        .ok_or_else(|| {
            SkypydbError::validation("Missing SALT_KEY. Set it in environment or .env.local.")
        })?;

    // Decode salt from base64
    let salt = BASE64_STANDARD
        .decode(salt_encoded)
        .map_err(|error| SkypydbError::validation(format!("Invalid SALT_KEY: {error}")))?;

    //Create a new client with encryption field set
    let client = ReactiveClient::new(
        "./db/_generated/skypydb_secure_rust.db",
        Some(encryption_key),
        Some(salt),
        Some(vec!["message".to_string()]),
    )?;
    let schema = schema::build_schema()?;

    //Get or create tables from the schema
    let tables = client.get_or_create_tables(&schema)?;

    //Error handling for tables
    let success_table = tables
        .get("success")
        .ok_or_else(|| skypydb::SkypydbError::table_not_found("Table 'success' not found"))?;

    //Add data to the tables
    let inserted = success_table.add(json_map! {
        "component" => "AuthService",
        "action" => "login",
        "message" => "User logged in successfully",
        "user_id" => "user123",
    })?;
    println!("Inserted: {inserted:?}");

    //Search for data
    let results = success_table.search(
        None,
        json_map! {
            "user_id" => "user123",
        },
    )?;

    //Print results
    println!("Decrypted results:");
    for result in results {
        println!("{result}");
    }

    Ok(())
}

File: demo/examples/python/secure_implementation/secure_example.py
import os
import skypydb
from dotenv import load_dotenv

# Load environment variables from .env file
load_dotenv(".env.local")

# Load encryption key from environment
encryption_key = os.getenv("ENCRYPTION_KEY") # create a encryption key and make it available in .env file before using it, don't show this key to anyone
salt_key = os.getenv("SALT_KEY") # create a salt key and make it available in .env file before using it, don't show this salt to anyone

# transform salt key to bytes
if salt_key is None:
    raise ValueError("SALT_KEY missing")
salt_bytes = salt_key.encode("utf-8")

# Create encrypted database
client = skypydb.ReactiveClient(
    encryption_key=encryption_key,
    salt=salt_bytes,
    encrypted_fields=["message"]  # Optional: encrypt only sensitive fields
)

# All operations work the same - encryption is transparent!
tables = client.get_or_create_table()
# if the tables already exists the programe get them instead

# Access your tables
success_table = tables["success"]
warning_table = tables["warning"]
error_table = tables["error"]

# Automatically encrypted
success_table.add(
    component="AuthService",
    action="login",
    message="User logged in successfully", # only this field is encrypted if encrypted_fields is not None
    user_id="user123"
)

# Data is automatically decrypted when retrieved
user_success_logs = success_table.search(
    user_id="user123"
)

if not user_success_logs:
    print("No results found.")
else:
    for user_success_log in user_success_logs:
        print(user_success_log)
